### PR TITLE
Create slider component

### LINKED ISSUE #5: Make the animation speed control in the UI a slider.
We can make the slider a range of -2 to 2 where negative number will reverse animation.

Screen Shot 2022-04-23 at 10 55 54 PM

These are some really cool ideas for slider controllers

https://twitter.com/anatudor/status/1501122828179066882?s=20&t=jvDZpoJpeQvmbiYuBGd1Vg

### PR DESCRIPTION
**Pull request by jandrade**
- Created `Slider` UI component using the `input[type=range]` element with custom styles.
- Replaced existing `speed` control with the new `Slider` component.
- Modified a React eslint rule to turn off the `PropTypes` rule.
Screen Shot 2022-05-01 at 4 39 53 PM

Closes #5

### CODE REVIEW

File: editor/app/components/slider.css [hunk 1]
```diff
@@ -0,0 +1,140 @@
+/**
+ * Slider component
+ */
+.slider {
+  --hue: 240;
```
Comment (jandrade) [inline editor/app/components/slider.css:5]: @MiguelCastillo I've created a bunch of CSS variables here, but if you think it's a bit overhead, I can remove them.

File: editor/app/components/slider.jsx [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import * as React from "react";
+import "./slider.css";
+
+/**
+ *
+ * @param {number} start - The start value of the range.
+ * @param {number} end - The end value of the range.
+ * @param {number} step - The increment given to each step.
```
Comment (MiguelCastillo) [inline editor/app/components/slider.jsx:8]: Thank you for these docs!!  Im having a lil hard time understanding `step`.

So if you have a range like 1 - 10, and step is 1, that means we are going to split 1-10 into chunks of 1 unit?  So we end up with `[1,2,3,4,5,6,7,8,9,10]`?

And if step is 2 then you end up with `[1, 3, 5, 7, 9]` or something like that?

Do you anticipate us using `step` for any value other than 1?   If not - maybe we can skip that as an option for now and we can add it when the tie comes?

Im fine landing like this, but curious what you think about adding a couple of unit tests, if nothing to test the two examples I provided above, and maybe one with a negative min.

**Review by MiguelCastillo — APPROVED** [review]
This is really great!!  Im excited to try it out and start using it other parts.

I left some comments.  Nothing blocking, just curious about your thoughts.

File: editor/app/components/slider.jsx [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+export default class Slider extends React.Component {
+  labels = [];
+
+  constructor(props) {
+    super(props);
+    const {min, max, step} = this.props;
+
+    // Create an array of labels for the slider here so these don't have to be
+    // recalculated on every render.
+    this.labels = range(min, max, step);
```
Comment (MiguelCastillo) [inline editor/app/components/slider.jsx:26]: oh I was thinking that we can convert this component to be a render function component instead of a class, and use the memo hook to calculate this labels values only once.  let me know what you think.

File: editor/app/components/slider.css [hunk 4]
```diff
@@ -0,0 +1,140 @@
...
+.slider-range,
+.slider-range::-webkit-slider-runnable-track,
+.slider-range::-webkit-slider-thumb {
+  box-sizing: border-box;
+  -webkit-appearance: none;
+  appearance: none;
+  background: transparent;
+}
+
+.slider-range::-webkit-slider-thumb {
```
[the remark this answers is not in the document]
Comment (jandrade) [reply editor/app/components/slider.css:92]: This is one "issue" due to browser compatibility... I tried to reuse the same rules in the same selector, but turns out that browsers don't allow that (at least for this selector) so I had to duplicate the styles into separate selectors.

That said, do you think we should include FF? or are we supporting webkit for now?

File: editor/app/components/slider.jsx [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+import "./slider.css";
+
+/**
+ *
+ * @param {number} start - The start value of the range.
+ * @param {number} end - The end value of the range.
+ * @param {number} step - The increment given to each step.
+ * @returns An array of numbers from start to end with increments based on step
+ */
+const range = (start, end, step) =>
```
[the remark this answers is not in the document]
Comment (MiguelCastillo) [reply editor/app/components/slider.jsx:11]: BTW - I dont think we really need tests if we removed steps...

File: editor/app/components/slider.jsx [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            </option>
+          ))}
+        </datalist>
+      </div>
+    );
+  }
+}
+
+Slider.defaultProps = {
+  step: 1,
```
[the remark this answers is not in the document]
Comment (jandrade) [reply editor/app/components/slider.jsx:61]: Also, now that you have converted some stuff over to hooks, I can probably do the same here and the result will be even better!

File: editor/app/components/slider.jsx (hunk 5 above)
[the remark this answers is not in the document]
Comment (jandrade) [reply editor/app/components/slider.jsx:11]: Gotcha! the only case I can think for now is if we needed/wanted to support smaller steps (e.g. `0.25, 0.5, 0.75....`). If we are ok with `1` for now, then definitively we can get rid of that argument.

File: editor/app/components/slider.jsx (hunk 3 above)
Comment (jandrade) [reply editor/app/components/slider.jsx:26]: ha! good call... I was just mentioning this in a previous comment, will do that in a moment.

File: editor/app/components/slider.css (hunk 4 above)
[the remark this answers is not in the document]
Comment (MiguelCastillo) [reply editor/app/components/slider.css:92]: Yeah let's include FF.  We should have a quick comment mentioning what you just said.  This also makes me think that we need postcss autoprefix if we end up having lots of vendor specific stuff.

**Comment by MiguelCastillo** [comment]
As you noticed when you were testing the slider and the old numeric control... but increasing the speed value causes jumps in the animation.  I am going to add some logic to smooth that out once this lands.  Question - does the controller give a value when you release the slider or does it generate values as you move the slider?

File: editor/app/components/slider.jsx [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+ * @param {number} start - The start value of the range.
+ * @param {number} end - The end value of the range.
+ * @returns An array of numbers from start to end.
+ */
+const range = (start, end) => Array.from({length: end - start + 1}, (_, i) => start + i);
+
+/**
+ * A custom slider component that extends the standard input range control.
+ */
+export default function Slider(props) {
```
Comment (MiguelCastillo) [inline editor/app/components/slider.jsx:16]: So much cleaner!

File: editor/app/scene/animation-properties.jsx [hunk 8]
```diff
@@ -58,15 +59,18 @@ export class AnimationProperties extends WithNodeState {
...
-          <input
-            type="number"
-            step=".1"
-            onChange={this.handletAnimationSpeedChange}
+          <Slider
             value={animationState.speed}
+            min={-1}
+            max={3}
+            onChange={this.handleAnimationSpeedChange}
+            step={1}
```
Comment (MiguelCastillo) [inline editor/app/scene/animation-properties.jsx:70]: this `step` seems to be leftover?

File: editor/app/components/slider.jsx (hunk 7 above)
Comment (jandrade) [reply editor/app/components/slider.jsx:16]: 🥳  100%

File: editor/app/scene/animation-properties.jsx (hunk 8 above)
Comment (jandrade) [reply editor/app/scene/animation-properties.jsx:70]: That's required by the `input` range element (forgot to mention that detail 😅): https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/range#step

Comment (jandrade) [reply editor/app/scene/animation-properties.jsx:70]: 🤔 that makes me think that  if we change the `step` value here, then the `range` fn will be disconnected from this value. We can solve this with two different approaches:

1. Adding back `step` to `range()`, so both the `input` and the `labels` are in sync.
2. Hard-code the `step` value in the `input[range]` component.

I'm slightly inclined for opt. 2, and as I said, if we need to change that later, then we can implement opt. 1.

What do you think?

Comment (MiguelCastillo) [reply editor/app/scene/animation-properties.jsx:70]: Sorry for the confusion. Yeah I was confused.  i thought that by removing the step in the `range` generation function we were removing the use of `step`.  But it seems like were using steps for two purposes.

1. for creating the range for the labels.
2. controlling the granularity in the input control.

I think what you have is great as is!!  We can remove it if you wanted to since the default value is already `1` in the component but it doesnt matter much at all.

Comment (jandrade) [reply editor/app/scene/animation-properties.jsx:70]: Removing `steps` for now, and will include it if we ever need it.

Comment (MiguelCastillo) [reply editor/app/scene/animation-properties.jsx:70]: actually - i never really asked you.  do you prefer keeping `step` everywhere the way you had it?

**Comment by jandrade** [comment]
> As you noticed when you were testing the slider and the old numeric control... but increasing the speed value causes jumps in the animation. I am going to add some logic to smooth that out once this lands. Question - does the controller give a value when you release the slider or does it generate values as you move the slider?

To answer your question... the controller generate values as I move the slider.
